src/python/pants/backend/go/target_type_rules.py

```diff
@@ -119,27 +151,63 @@ async def infer_go_dependencies(
             inferred_dependencies.append(first_party_import_path_to_address[import_path])
             continue
 
-        # Infer third-party dependencies on go_external_module targets.
-        found_module_import = False
-        for module_import_path, address in third_party_import_path_to_address.items():
-            # TODO: This check assumes that the import path for the external module is a prefix of any package
-            # from the external module and that the import path will never overlap with another external module's
-            # import path. This may be the case, but bears confirmation.
-            #
-            # TODO: Also cleanup mandatory go_module ownership, so module_import_path is not Optional[str].
-            if module_import_path and import_path.startswith(module_import_path):
-                inferred_dependencies.append(address)
-                found_module_import = True
-
-        if found_module_import:
+        # Infer third-party dependencies on _go_ext_mod_package targets.
+        candidate_third_party_packages = package_mapping.mapping.get(import_path, ())
+        if len(candidate_third_party_packages) > 1:
+            logger.warning(
+                f"Ambiguous mapping for import path {import_path} on packages at addresses: {candidate_third_party_packages}"
+            )
```

Comment:
This should be using `ExplicitlyProvidedDependencies.maybe_warn_of_ambiguous_dependency_inference()` and `ExplicitlyProvidedDependencies.disambiguated()`. Please add a TODO if you want to punt on that.

Reply:
TODO